File: 2022/Day_04/main.hs
import Data.Functor ((<&>))
import Control.Arrow ((>>>))
import Data.List (intersect)
import Data.List.Split ( splitOn )

parseInput :: String -> IO [([Int],[Int])]
parseInput filename =
    readFile filename
    <&> (lines
    >>> map (
        splitOn "," 
        >>> map (
            splitOn "-" 
            >>> (\(a:b:_) -> [read a..read b])) 
        >>> (\(a:b:_) -> (a,b)))
    )

part1 :: [([Int],[Int])] -> Int
part1 = foldr (\(elf1, elf2) acc ->
    if elf1 `intersect` elf2 == elf2 || elf1 `intersect` elf2 == elf1 then
        acc + 1
    else
        acc
    ) 0

part2 :: [([Int],[Int])] -> Int
part2 = foldr (\(elf1, elf2) acc ->
    if not $ null $ elf1 `intersect` elf2 then
        acc + 1
    else
        acc
    ) 0

main = do
    parsedInput <- parseInput "input.txt"
    putStrLn $ "Part 1: " ++ show (part1 parsedInput)
    putStrLn $ "Part 2: " ++ show (part2 parsedInput)

File: 2022/Day_04/main.exs
defmodule Day_04 do

  def parse_input(filename) do
    File.read!(filename)
    |> String.split("\n", trim: true)
    |> Enum.map(fn line ->
      Regex.run(~r/(\d+)-(\d+),(\d+)-(\d+)/, line)
      |> tl
      |> Enum.map(&String.to_integer/1)
      |> then(fn [elf1_i, elf1_f, elf2_i, elf2_f] ->
        {MapSet.new(elf1_i..elf1_f), MapSet.new(elf2_i..elf2_f)}
      end)
    end)
  end

  def part1(input) do
    input
    |> Enum.reduce(0, fn {elf1, elf2}, acc ->
      if MapSet.subset?(elf1, elf2) or MapSet.subset?(elf2, elf1) do
        acc + 1
      else
        acc
      end
    end)
  end

  def part2(input) do
    input
    |> Enum.reduce(0, fn {elf1, elf2}, acc ->
      if MapSet.intersection(elf1, elf2) |> MapSet.size > 0 do
        acc + 1
      else
        acc
      end
    end)
  end

  def main do
    filename = Enum.at(System.argv, 0, "input.txt")
    parsed_input = parse_input(filename)

    IO.inspect(part1(parsed_input), label: "Part 1")
    IO.inspect(part2(parsed_input), label: "Part 2")
  end
end

Day_04.main
